### PR TITLE
PLT-1095 shutdown hook

### PR DESCRIPTION
**Pull request by yarikos**
elasticio/api#431
elasticio/api#432

### CODE REVIEW

**Comment by drobiazko** [comment]
@yarikos please don't merge it until we want to release it. There is [another pull request](https://github.com/elasticio/sailor-nodejs/pull/46) waiting to become part of v2.1.0 release together with that one.

File: lib/hooksData.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        const options = {
+            url: `${this.basePath}/sailor-support/hooks/task/${this.taskId}/startup/data`,
+            method,
+            auth: {
+                user: this.user,
+                pass: this.pass
+            }
+        };
+        if (data) {
+            options.form = data;
```
Comment (drobiazko) [inline lib/hooksData.js]: Why not ``options.form``?

File: lib/hooksData.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        }
+        console.log('\n\n\n options', options, '\n\n\n');
```
Comment (drobiazko) [inline lib/hooksData.js]: Remove this log

File: lib/sailor.js [hunk 3]
```diff
@@ -63,15 +64,37 @@ class Sailor {
...
             console.log('Starting up component');
             const result = yield this.invokeModuleFunction('startup');
-            //TODO: persist it
+
+            const handle = hooksData.startup(this.settings);
+            try {
+                yield handle.create(result);
+            } catch (e) {
+                //@todo check only http 409
+                if (0) {
```
Comment (drobiazko) [inline lib/sailor.js]: ???

File: lib/hooksData.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+class HooksData {
+
+    constructor(taskId, user, pass) {
+        this.taskId = taskId;
+        this.user = user;
+        this.pass = pass;
+        this.basePath = process.env.ELASTICIO_API_URI || 'https://api.elastic.io/';
+    }
+
+    request(method, data) {
```
Comment (drobiazko) [inline lib/hooksData.js]: If ``data`` is not defined or is an empty json object, we can print a warning a return here immdiatelly

File: lib/sailor.js [hunk 5]
```diff
@@ -63,15 +64,37 @@ class Sailor {
...
 
     shutdown() {
         return co(function* doShutdown() {
             console.log('Shutting down component');
-            return yield this.invokeModuleFunction('shutdown');
+            let startupData;
+            try {
+                const handle = hooksData.startup(this.settings);
+                startupData = yield handle.retrieve();
+                console.log('startupData', startupData);
```
Comment (drobiazko) [inline lib/sailor.js]: Don't forget to remove this log

File: lib/sailor.js [hunk 6]
```diff
@@ -63,15 +64,37 @@ class Sailor {
...
+            } catch (e) {
+                console.log('startupData fail', e);
+                if (!e === 404) {
+                    //@todo log errors if not 404 error
```
Comment (drobiazko) [inline lib/sailor.js]: ??

File: mocha_spec/run.spec.js [hunk 7]
```diff
@@ -227,15 +231,18 @@ describe('Integration Test', () => {
...
             expect(req.body).to.eql({
                 url: 'https://in.elastic.io/hooks/5559edd38968ec0736000003'
             });
-            res.json({
-                id: 'webhook_123'
-            });
+            res.json(subsriptionResponse);
         });
 
+        // sailor retrievs info about current step from API
```
Comment (drobiazko) [inline mocha_spec/run.spec.js]: typof in _retrievs_. it should be _retrieves_

File: run.js [hunk 8]
```diff
@@ -6,14 +6,30 @@ const co = require('co');
...
+let disconnectRequired;
 
 co(function* putOutToSea() {
     sailor = new Sailor(settings);
+
+    if (!!settings.HOOK_SHUTDOWN) {
+        disconnectRequired = false;
+        sailor.reportError = () => {
+        }
+        console.log('HOOK_SHUTDOWN')
```
Comment (drobiazko) [inline run.js]: Please provide nice log statements

File: lib/hooksData.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+
+const rp = require('request-promise-native')
+
+class HooksData {
+
+    constructor(taskId, user, pass) {
+        this.taskId = taskId;
+        this.user = user;
+        this.pass = pass;
+        this.basePath = process.env.ELASTICIO_API_URI || 'https://api.elastic.io/';
```
Comment (drobiazko) [inline lib/hooksData.js]: The last slash must be removed due to the way the url is constructed down there.

File: lib/sailor.js [hunk 10]
```diff
@@ -61,15 +62,32 @@ class Sailor {
         return co(function* doStartup() {
             console.log('Starting up component');
             const result = yield this.invokeModuleFunction('startup');
-            //TODO: persist it
+
+            const handle = hooksData.startup(this.settings);
+            try {
+                const dataToStore = _.isEmpty(result) ? {} : result;
```
Comment (drobiazko) [inline lib/sailor.js]: Let's call it ``state``

File: lib/sailor.js [hunk 11]
```diff
@@ -61,15 +62,32 @@ class Sailor {
...
+                yield handle.create(dataToStore);
+            } catch (e) {
+                if (e.statusCode === 409) {
+                    console.warn('Hooks data creating conflict. Deleting existing data');
+                    yield handle.delete();
+                    yield handle.create(result);
```
Comment (drobiazko) [inline lib/sailor.js]: What happens if this fails?

File: lib/sailor.js [hunk 12]
```diff
@@ -61,15 +62,32 @@ class Sailor {
...
+
             return result;
         }.bind(this));
     }
 
     shutdown() {
         return co(function* doShutdown() {
             console.log('Shutting down component');
-            return yield this.invokeModuleFunction('shutdown');
+            let startupData;
```
Comment (drobiazko) [inline lib/sailor.js]: Let's call it ``state``

File: lib/sailor.js [hunk 13]
```diff
@@ -61,15 +62,32 @@ class Sailor {
...
+            const handle = hooksData.startup(this.settings);
+            startupData = yield handle.retrieve();
+            yield this.invokeModuleFunction('shutdown', startupData);
+            yield handle.delete();
```
Comment (drobiazko) [inline lib/sailor.js]: What happens if this fails?

File: lib/sailor.js (hunk 11 above)
Comment (yarikos) [inline lib/sailor.js]: Fail will cause an exception, which will be caught by general error handler and sailor will exit with log message about error.

File: lib/sailor.js [hunk 14]
```diff
@@ -96,8 +96,9 @@ class Sailor {
     init() {
         return co(function* doInit() {
             log.info('hook-init start');
-            yield this.invokeModuleFunction('init');
+            const res = yield this.invokeModuleFunction('init');
             log.info('hook-init finished');
```
Comment (drobiazko) [inline lib/sailor.js]: Finished initialisation.

File: lib/sailor.js [hunk 15]
```diff
@@ -60,41 +60,44 @@ class Sailor {
...
-
+            log.info('hook-startup finished');
             return result;
         }.bind(this));
     }
 
     shutdown() {
         return co(function* doShutdown() {
-            console.log('Shutting down component');
+            log.info('hook-shutdown begin');
```
Comment (drobiazko) [inline lib/sailor.js]: About to shutdown

File: lib/sailor.js [hunk 16]
```diff
@@ -60,41 +60,44 @@ class Sailor {
...
             const handle = hooksData.startup(this.settings);
             const state = yield handle.retrieve();
             yield this.invokeModuleFunction('shutdown', state);
             yield handle.delete();
+            log.info('hook-shutdown finished');
```
Comment (drobiazko) [inline lib/sailor.js]: Shut down successfully

File: lib/sailor.js [hunk 17]
```diff
@@ -60,41 +60,44 @@ class Sailor {
...
         }.bind(this));
     }
 
     init() {
         return co(function* doInit() {
-            console.log('Initializing component for execution');
-            return yield this.invokeModuleFunction('init');
+            log.info('hook-init start');
```
Comment (drobiazko) [inline lib/sailor.js]: About to Initialise component for execution

File: lib/sailor.js [hunk 18]
```diff
@@ -60,41 +60,44 @@ class Sailor {
...
                 const state = _.isEmpty(result) ? {} : result;
                 yield handle.create(state);
             } catch (e) {
                 if (e.statusCode === 409) {
-                    console.warn('Hooks data creating conflict. Deleting existing data');
+                    log.warn('hook-startup – hooks data creating conflict. Deleting existing data');
                     yield handle.delete();
                     yield handle.create(result);
                 } else {
+                    log.warn('hook-startup – error');
```
Comment (drobiazko) [inline lib/sailor.js]: What the point of this? Isn't the error reported anyway?

File: lib/sailor.js [hunk 19]
```diff
@@ -60,41 +60,44 @@ class Sailor {
 
     startup() {
         return co(function* doStartup() {
-            console.log('Starting up component');
+            log.info('hook-startup begin');
             const result = yield this.invokeModuleFunction('startup');
```
Comment (drobiazko) [inline lib/sailor.js]: Why rewriting the user friendly message to something cryptical?

File: lib/sailor.js [hunk 20]
```diff
@@ -60,41 +60,44 @@ class Sailor {
...
-
+            log.debug('hook-startup result', { result });
```
Comment (drobiazko) [inline lib/sailor.js]: Good idea? A developer can do it in his component code.

File: lib/sailor.js [hunk 21]
```diff
@@ -60,41 +60,44 @@ class Sailor {
...
             const handle = hooksData.startup(this.settings);
             try {
                 const state = _.isEmpty(result) ? {} : result;
                 yield handle.create(state);
             } catch (e) {
                 if (e.statusCode === 409) {
-                    console.warn('Hooks data creating conflict. Deleting existing data');
+                    log.warn('hook-startup – hooks data creating conflict. Deleting existing data');
```
Comment (drobiazko) [inline lib/sailor.js]: Use something more user friendly

File: lib/sailor.js (hunk 20 above)
Comment (yarikos) [inline lib/sailor.js]: It's debug, by default it's hidden

File: lib/sailor.js (hunk 18 above)
Comment (yarikos) [inline lib/sailor.js]: It's reported, general error handler doesn't tell suggestion where exactly error has occurred (e.g. when stacktrace point to somewhere in request.js etc).

File: run.js [hunk 22]
```diff
@@ -6,14 +6,29 @@ const co = require('co');
...
 let sailor;
+let disconnectRequired;
 
 co(function* putOutToSea() {
     sailor = new Sailor(settings);
+
+
+    if (!!settings.HOOK_SHUTDOWN) {
+        disconnectRequired = false;
+        sailor.reportError = () => {
```
Comment (drobiazko) [inline run.js]: Just out of curiosity. Why?

Comment (yarikos) [inline run.js]: `reportError` sends some data into `amqp`.
In `shutdown` mode sailor doesn't listen to `amqp`.
So this workaround is in order to disable error reporting.

**Pull request merged**